src/irmin-pack/mem/content_addressable.ml

```diff
@@ -0,0 +1,148 @@
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open! Import
+
+module Pool : sig
+  type ('k, 'v) t
+  (** Reference-counted pool of values with corresponding keys. *)
+
+  val create : alloc:('k -> 'v) -> ('k, 'v) t
+  (** Get an empty pool, given a function for allocating new instances from IDs. *)
+
+  val take : ('k, 'v) t -> 'k -> 'v
+  (** Get an instance from the pool by its key, allocating it if necessary. *)
+
+  val drop : ('k, 'v) t -> 'k -> unit
+  (** Reduce the reference count of an element, discarding it if the reference
+      count drops to 0. *)
+end = struct
+  type 'v elt = { mutable refcount : int; instance : 'v }
+  type ('k, 'v) t = { instances : ('k, 'v elt) Hashtbl.t; alloc : 'k -> 'v }
+
+  let create ~alloc = { instances = Hashtbl.create 0; alloc }
+
+  let take t k =
+    match Hashtbl.find_opt t.instances k with
+    | Some elt ->
+        elt.refcount <- succ elt.refcount;
+        elt.instance
+    | None ->
+        let instance = t.alloc k in
+        Hashtbl.add t.instances k { instance; refcount = 1 };
+        instance
+
+  let drop t k =
+    match Hashtbl.find_opt t.instances k with
+    | None -> failwith "Pool.drop: double free"
+    | Some { refcount; _ } when refcount <= 0 -> assert false
+    | Some { refcount = 1; _ } -> Hashtbl.remove t.instances k
+    | Some elt -> elt.refcount <- pred elt.refcount
+end
+
+module Maker (K : Irmin.Hash.S) = struct
+  type key = K.t
+
+  module Make (Val : Irmin_pack.Pack_value.S with type hash := K.t) = struct
+    module KMap = Map.Make (struct
+      type t = K.t
+
+      let compare = Irmin.Type.(unstage (compare K.t))
+    end)
+
+    type key = K.t
+    type value = Val.t
+
+    type 'a t = {
+      name : string;
+      mutable t : value KMap.t;
+      mutable generation : int63;
+    }
+
+    let instances =
+      Pool.create ~alloc:(fun name ->
+          { name; t = KMap.empty; generation = Int63.zero })
+
+    let v name = Lwt.return (Pool.take instances name)
+    let equal_key = Irmin.Type.(unstage (equal K.t))
+
+    let clear_keep_generation t =
+      Log.debug (fun f -> f "clear_keep_generation");
+      t.t <- KMap.empty;
+      Lwt.return_unit
+
+    let clear t =
+      Log.debug (fun f -> f "clear");
+      t.t <- KMap.empty;
+      t.generation <- Int63.succ t.generation;
+      Lwt.return_unit
+
+    let close t =
+      Log.debug (fun f -> f "close");
+      Pool.drop instances t.name;
```

Comment:
Should we attach a finaliser somewhere in order to drop the instance when the user lost the reference to the store without closing it?

Reply:
Perhaps we should, but if so I think it's a good idea to that in the regular `irmin-pack` case for consistency.

(I really wish the OCaml stdlib had a `std::shared_ptr`-like abstraction for this...)

Reply:
What do you think about ephemeron?

Reply:
Indeed, I think https://ocaml.org/api/Ephemeron.GenHashTable.html is close to providing what we want (and maybe 100% of what we want). I think it doesn't allow the user to specify a tear-down function for "dead" bindings, but maybe I should play with it and find out.

(Unfortunately, ephemerons aren't portable to JS.)